files/aws_lifecycle_hooks/tools.py

```diff
@@ -74,19 +116,128 @@ def get_asg_name(
         instance_id: str,
         asg_client,
 ) -> str:
-    asg_info = asg_client.describe_auto_scaling_instances(InstanceIds=[instance_id])
-    asg_name = asg_info[u'AutoScalingInstances'][0][u'AutoScalingGroupName']
-    print("I am part of Auto Scaling Group {asg_name}".format(asg_name=asg_name))
+    """Use describe asg to get asg group name."""
+    import botocore.exceptions
+    asg_name = None
+    try:
+        asg_info = asg_client.describe_auto_scaling_instances(InstanceIds=[instance_id])
+        asg_name = asg_info[u'AutoScalingInstances'][0][u'AutoScalingGroupName']
+        print("I am part of Auto Scaling Group {asg_name}".format(asg_name=asg_name))
+    except botocore.exceptions.ClientError as e:
+        if e.response['Error']['Code'] == "AccessDenied":
+            print('This instance/user has no permission to query autoscaling.')
+        else:
+            raise
 
     return asg_name
 
 
+@functools.lru_cache(maxsize=1)
+def get_block_device_mapping() -> list:
+    """Get meta-data Block device mapping."""
+    block_device_mapping = get_metadata('meta-data/block-device-mapping')
+    print("My raw block-device-mapping is {block_device_mapping}".format(block_device_mapping=block_device_mapping))
+    block_device_mapping_list = [x for x in block_device_mapping.splitlines()]
+    return block_device_mapping_list
+
+
+def get_block_device_mapping_filtered(regex: str) -> list:
```

Comment:
I would require a `re.Pattern` here, and push the re.compile() to the caller. This "guarantees" that you get a valid regex.